sky/skylet/job_lib.py

```diff
@@ -783,26 +805,34 @@ def cancel_jobs_encoded_results(jobs: Optional[List[int]],
             # Cancel jobs with specified IDs.
             job_records = _get_jobs_by_ids(jobs)
 
-    # TODO(zhwu): `job_client.stop_job` will wait for the jobs to be killed, but
-    # when the memory is not enough, this will keep waiting.
-    job_client = _create_ray_job_submission_client()
     cancelled_ids = []
 
     # Sequentially cancel the jobs to avoid the resource number bug caused by
     # ray cluster (tracked in #1262).
-    for job in job_records:
-        job_id = make_ray_job_id(job['job_id'])
+    for job_record in job_records:
+        job_id = job_record['job_id']
         # Job is locked to ensure that pending queue does not start it while
         # it is being cancelled
-        with filelock.FileLock(_get_lock_path(job['job_id'])):
-            try:
-                job_client.stop_job(job_id)
-            except RuntimeError as e:
-                # If the request to the job server fails, we should not
-                # set the job to CANCELLED.
-                if 'does not exist' not in str(e):
-                    logger.warning(str(e))
-                    continue
+        with filelock.FileLock(_get_lock_path(job_id)):
+            job = _get_jobs_by_ids([job_id])[0]
+            if job['pid'] > 0:
+                # Not use process.terminate() as that will only terminate the
+                # process shell process, not the ray driver process
+                # under the shell.
+                # No need to use `subprocess_utils.kill_children_processes`,
+                # as the underlying ray driver process will handle the
+                # termination of user programs (we use SIGTERM to allow ray
+                # driver to gracefully exit).
+                try:
+                    os.killpg(job['pid'], signal.SIGTERM)
```

Comment:
Since this is a `SIGTERM`, this may potentially block on/leak misbehaving processes. Should we have a check for processes that failed to kill after a fixed timeout, and use `SIGKILL` for those misbehaving processes?

Reply:
This is a great catch! I replaced it with our `kill_children_processes` implementation, which is adopted from another on-going branch. : )

Reply:
Actually, it is more efficient and robust to use our process killing daemon to handle the asynchronous process killing, switched that implementation